File: Task_2_SensorReader/app/src/main/java/com/example/sensorreader/MainActivity.java
package com.example.sensorreader;

import android.hardware.Sensor;
import android.os.Bundle;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.sensorreader.adapter.SensorAdapter;
import com.example.sensorreader.manager.DeviceSensorManager;
import com.example.sensorreader.model.SensorItem;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MainActivity extends AppCompatActivity
        implements DeviceSensorManager.OnSensorDataListener {

    private RecyclerView rvSensors;
    private DeviceSensorManager sensorManager;
    private SensorAdapter sensorAdapter;
    private List<SensorItem> sensorItemList;

    private Map<Integer, Integer> sensorTypeToIndexMap;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();
        initSensorManager();
        buildSensorList();
        setupRecyclerView();
    }


    private void initViews() {
        rvSensors = findViewById(R.id.rvSensors);
        TextView tvDeviceInfo = findViewById(R.id.tvDeviceInfo);
        String deviceDescription = "Device: " + android.os.Build.MODEL +
                " | Android " + android.os.Build.VERSION.RELEASE;
        tvDeviceInfo.setText(deviceDescription);
    }

    private void initSensorManager() {
        sensorManager = new DeviceSensorManager(this);
        sensorManager.setOnSensorDataListener(this);
    }

    private void buildSensorList() {
        sensorItemList = new ArrayList<>();
        sensorTypeToIndexMap = new HashMap<>();

        for (int i = 0; i < DeviceSensorManager.TARGET_SENSORS.length; i++) {
            DeviceSensorManager.SensorConfig config = DeviceSensorManager.TARGET_SENSORS[i];
            boolean available = sensorManager.isSensorAvailable(config.type);

            SensorItem item = new SensorItem(
                    config.name,
                    getSensorTypeLabel(config.type),
                    config.unit,
                    available
            );
            sensorItemList.add(item);
            sensorTypeToIndexMap.put(config.type, i);
        }
    }

    private void setupRecyclerView() {
        sensorAdapter = new SensorAdapter(sensorItemList);
        rvSensors.setLayoutManager(new LinearLayoutManager(this));
        rvSensors.setAdapter(sensorAdapter);
    }

    @Override
    public void onSensorDataUpdated(int sensorType, float[] values) {
        Integer index = sensorTypeToIndexMap.get(sensorType);
        if (index == null) return;

        String formattedValue = DeviceSensorManager.formatValues(sensorType, values);
        String unit = sensorItemList.get(index).getUnit();
        String displayText = formattedValue + "  " + unit;

        runOnUiThread(() -> sensorAdapter.updateSensorValue(index, displayText));
    }


    @Override
    protected void onResume() {
        super.onResume();
        // Start listening when app is visible
        sensorManager.startListening();
    }

    @Override
    protected void onPause() {
        super.onPause();
        // Stop listening when app goes to background — saves battery!
        sensorManager.stopListening();
    }

    private String getSensorTypeLabel(int type) {
        switch (type) {
            case Sensor.TYPE_ACCELEROMETER:
            case Sensor.TYPE_GYROSCOPE:
            case Sensor.TYPE_GRAVITY:
            case Sensor.TYPE_LINEAR_ACCELERATION:
                return "Motion";
            case Sensor.TYPE_MAGNETIC_FIELD:
            case Sensor.TYPE_LIGHT:
            case Sensor.TYPE_PRESSURE:
            case Sensor.TYPE_AMBIENT_TEMPERATURE:
            case Sensor.TYPE_RELATIVE_HUMIDITY:
                return "Environment";
            default:
                return "Other";
        }
    }
}

File: SensorReader/app/src/main/java/com/example/sensorreader/MainActivity.kt
package com.example.sensorreader

import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.Bundle
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class MainActivity : AppCompatActivity() {

    private lateinit var sensorManager: SensorManager
    private val sensorList = mutableListOf<SensorInfo>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager

        loadSensors()

        val tvCount = findViewById<TextView>(R.id.tvSensorCount)
        tvCount.text = getString(R.string.sensor_count_format, sensorList.size)

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = SensorAdapter(sensorList) { sensorInfo ->
            val intent = Intent(this, SensorDetailActivity::class.java).apply {
                putExtra(SensorDetailActivity.EXTRA_SENSOR_TYPE, sensorInfo.type)
                putExtra(SensorDetailActivity.EXTRA_SENSOR_NAME, sensorInfo.name)
            }
            startActivity(intent)
        }
    }

    private fun loadSensors() {
        val allSensors = sensorManager.getSensorList(Sensor.TYPE_ALL)
        for (sensor in allSensors) {
            sensorList.add(
                SensorInfo(
                    sensor = sensor,
                    name = sensor.name,
                    type = sensor.type,
                    vendor = sensor.vendor,
                    typeName = getSensorTypeName(sensor.type)
                )
            )
        }
    }

    private fun getSensorTypeName(type: Int): String = when (type) {
        Sensor.TYPE_ACCELEROMETER -> "Accelerometer"
        Sensor.TYPE_MAGNETIC_FIELD -> "Magnetic Field"
        Sensor.TYPE_GYROSCOPE -> "Gyroscope"
        Sensor.TYPE_LIGHT -> "Light"
        Sensor.TYPE_PRESSURE -> "Pressure"
        Sensor.TYPE_PROXIMITY -> "Proximity"
        Sensor.TYPE_GRAVITY -> "Gravity"
        Sensor.TYPE_LINEAR_ACCELERATION -> "Linear Acceleration"
        Sensor.TYPE_ROTATION_VECTOR -> "Rotation Vector"
        Sensor.TYPE_RELATIVE_HUMIDITY -> "Relative Humidity"
        Sensor.TYPE_AMBIENT_TEMPERATURE -> "Ambient Temperature"
        Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED -> "Magnetic Field (Uncalibrated)"
        Sensor.TYPE_GAME_ROTATION_VECTOR -> "Game Rotation Vector"
        Sensor.TYPE_GYROSCOPE_UNCALIBRATED -> "Gyroscope (Uncalibrated)"
        Sensor.TYPE_SIGNIFICANT_MOTION -> "Significant Motion"
        Sensor.TYPE_STEP_DETECTOR -> "Step Detector"
        Sensor.TYPE_STEP_COUNTER -> "Step Counter"
        Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR -> "Geomagnetic Rotation Vector"
        Sensor.TYPE_HEART_RATE -> "Heart Rate"
        Sensor.TYPE_POSE_6DOF -> "Pose 6DOF"
        Sensor.TYPE_STATIONARY_DETECT -> "Stationary Detect"
        Sensor.TYPE_MOTION_DETECT -> "Motion Detect"
        Sensor.TYPE_HEART_BEAT -> "Heart Beat"
        Sensor.TYPE_LOW_LATENCY_OFFBODY_DETECT -> "Low Latency Off-Body Detect"
        Sensor.TYPE_ACCELEROMETER_UNCALIBRATED -> "Accelerometer (Uncalibrated)"
        else -> "Type $type"
    }
}
